### ISSUE TITLE
rust-analyzer needs access to `$RUSTUP_HOME`.

### ISSUE THREAD

**Issue by WhyNotHugo** [lspcontainers/dockerfiles#55]
rust-analyzer attempts to access crates for the current project, which are usually stored in `~/.cache/cargo`. This should be mounted as a volume.

Without this directory:
- The language server needs network access (non-ideal).
- The language server will download and compile all dependencies on each run, which is extremely slow and consumes **a lot** of network.

This isn't the first LSP to require an additional volume for this kind of usage. I'm thinking of maybe adding an extra item to [objects in the supported_languages table](https://github.com/lspcontainers/lspcontainers.nvim/blob/main/lua/lspcontainers/init.lua#L15-L32) (e.g.: `extra_volumes`), with paths to additional volumes that they each require.

This provides a base onto which individual LSPs can specify extra volumes (and, probably, whether these volumes should be `:ro`or `:rw`).

**Comment by erikreinert** [comment]
I believe this is something we still need to solve for - right?

**Comment by WhyNotHugo** [comment]
Yup, still unresolved. I think adding an `extra_volumes` entry (with a `/path/on/host:/path/on/container` format) would cover this, and a few other situations quite well (e.g.: `golang` has a volume hardcoded which could use leverage this design).

**Comment by B3RR10** [comment]
I think this ticket belongs to the [lspcontainers.nvim](https://github.com/lspcontainers/lspcontainers.nvim) repo, right? My workaround was to define the `cmd_builder` similar to the default go command.

```
  cmd = require('lspcontainers').command('rust_analyzer', {
    network = 'default',
    cmd_builder = function(runtime, workdir, image, network)
      local volume = workdir .. ':' .. workdir
      local env = vim.api.nvim_eval('environ()')
      local cargo_path = env.HOME .. '/.cargo'
      local cargo_path_volume = cargo_path .. ':' .. '/root/.cargo'

      -- add ':z' to podman volumes to avoid permission denied errors
      if runtime == 'podman' then
        cargo_path_volume = cargo_path_volume .. ':z'
        volume = volume .. ':z'
      end

      return {
        runtime,
        'container',
        'run',
        '--interactive',
        '--network=' .. network,
        '--rm',
        '--workdir=' .. workdir,
        '--volume=' .. volume,
        '--volume=' .. cargo_path_volume,
        image,
      }
    end,
  }),
```
But I agree that an `extra_volumes` property would be a nice solution.

**Comment by WhyNotHugo** [comment]
You're snippet assumes a specific value for RUSTUP_HOME, it should read the environment variable instead.

I don't think it's the right approach anyway. We should add a new property to LSPs: a function that returns additional volumes. Otherwise each LSP with any volumes re-implements the whole builder function.

**Comment by B3RR10** [comment]
> You're snippet assumes a specific value for RUSTUP_HOME, it should read the environment variable instead. I don't think it's the right approach anyway. We should add a new property to LSPs: a function that returns additional volumes. Otherwise each LSP with any volumes re-implements the whole builder function.

I totally agree with you... It was just a workaround :)

**Comment by erikreinert** [comment]
Okay here is what I'm going to do now that I fully understand this. Two things:

- I want to re-write `lspcontainers` now that I have the knowledge I do about Lua and how these language servers can work
- I want to consider these types of problems at a higher level as suggested here

I think what I'm going to do is create a new `v2` issue with all of the features we would like to implement and improve upon. Then we start from scratch and make this even better.

The original code was really a prototype that we have leaned on but now that we want to support multiple backends, architectures, volumes, etc we need to solve these problems in clearer ways.

I will close this ticket for now and reference it in the new V2 issue to consider changes.

Thank you all for your continued support with this project. IT IS one I want to keep going but think we have places we need to properly consider to better maintain for the future.

**Comment by WhyNotHugo** [comment]
I moved to sandboxing several LSPs via bubblewrap, because that allows me to use the distribution-provided LSP. This was especially handy for LSPs with no existing docker image, but also allowed me to de-duplicate a lot of stuff locally and avoid having to build/pull images (especially on aarch64 or after nuking everything docker to clean up disk space)..

If you're going to re-write a lot of this, maybe the design of what I did can serve as a useful reference.:
https://git.sr.ht/~whynothugo/dotfiles/tree/2ca611de/item/home/.config/nvim/lua/lsp.lua#L43-153

Mostly, I added the sandboxing-related config params to the usual LSP configurations map under a `sandbox` key with a consistent format. I used a wrapper for the `setup()` call which applies this.

I set each LSP's `cmd` to `false`, and then add a wrapper for `config.on_new_config` which redefines the `cmd` when the config changes. This lets me update the flags for the LSP to include the root directory of the file being editing (which might change after vim's startup). It's also a great spot to set `params.processId = vim.NIL` for all sandboxes LSPs, since this interferes with LSPs that don't run on the same PID namespace as neovim.

Anyway, feel free to poke around if you wish.
